### PR TITLE
WT-7723 Delete the updates in the history store if they are rolled back or is the first stable update on the update chain 

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/txn/txn_rollback_to_stable.c [hunk 1]
```diff
@@ -791,22 +863,24 @@ __rollback_abort_col_var(WT_SESSION_IMPL *session, WT_REF *ref, wt_timestamp_t r
...
-static void
+static int
 __rollback_abort_col_fix(WT_SESSION_IMPL *session, WT_PAGE *page, wt_timestamp_t rollback_timestamp)
 {
     WT_INSERT_HEAD *ins;
 
     /* Review the changes to the original on-page data items */
     if ((ins = WT_COL_UPDATE_SINGLE(page)) != NULL)
-        __rollback_abort_insert_list(session, ins, rollback_timestamp);
+        WT_RET(__rollback_abort_insert_list(session, page, ins, rollback_timestamp, NULL));
```
Comment (kommiharibabu) [inline src/txn/txn_rollback_to_stable.c:873]: For fixed-length column store type, there is no history store used. Do we need to either don't pass the page here or add a check in the location to not allocate memory and etc?

File: src/txn/txn_rollback_to_stable.c [hunk 2]
```diff
@@ -940,42 +1016,50 @@ __rollback_abort_row_leaf(WT_SESSION_IMPL *session, WT_REF *ref, wt_timestamp_t
...
-            WT_RET(__rollback_abort_ondisk_kv(session, ref, NULL, rip, rollback_timestamp, 0));
+            WT_ERR(__rollback_abort_ondisk_kv(session, ref, NULL, rip, rollback_timestamp, 0));
     }
 
     /*
      * If the configuration is not in-memory, abort history store updates from the reconciled pages
      * of data store.
      */
     if (!F_ISSET(S2C(session), WT_CONN_IN_MEMORY))
-        WT_RET(__rollback_abort_row_reconciled_page(session, page, rollback_timestamp));
```
Comment (kommiharibabu) [inline src/txn/txn_rollback_to_stable.c:973]: There is no need for this function now. This is to remove the additional contents that went into the history store.

File: src/txn/txn_rollback_to_stable.c [hunk 3]
```diff
@@ -15,37 +15,80 @@
...
+    hs_cursor->set_key(hs_cursor, 4, S2BT(session)->id, key, WT_TS_MAX, UINT64_MAX);
+    ret = __wt_curhs_search_near_before(session, hs_cursor);
+    for (; ret == 0; ret = hs_cursor->prev(hs_cursor)) {
+        WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, hs_key, &hs_start_ts, &hs_counter));
+        if (hs_start_ts < ts)
+            break;
+        WT_ERR(hs_cursor->remove(hs_cursor));
+        WT_STAT_CONN_DATA_INCR(session, txn_rts_hs_removed);
+        if (hs_start_ts == ts)
+            WT_STAT_CONN_DATA_INCR(session, cache_hs_key_truncate_rts_unstable);
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/txn/txn_rollback_to_stable.c:59]: It's just to be consistent with the behaviour in the history store rollback code.

File: src/txn/txn_rollback_to_stable.c (hunk 1 above)
Comment (quchenhao) [reply src/txn/txn_rollback_to_stable.c:873]: We only allocate an u64 integer so I think it is fine. We pay another price if we do the type check.

File: src/txn/txn_rollback_to_stable.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c:59]: The history store rollback code doesn't read an extra key. whatever key it reads it removes it. In this case, it doesn't.

File: src/txn/txn_rollback_to_stable.c (hunk 1 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c:873]: ok, undertood.

**Comment by kommiharibabu** [comment]
@quchenhao __rollback_ondisk_fixup_key() is passed with false flag code is removed in the above commit, so there is no need for the code to handle for false.

File: src/txn/txn_rollback_to_stable.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/txn/txn_rollback_to_stable.c:59]: No, it removes the restored key as well. Here is the same.

**Comment by quchenhao** [comment]
> @quchenhao __rollback_ondisk_fixup_key() is passed with false flag code is removed in the above commit, so there is no need for the code to handle for false.

I don't understand what this means.

**Review by kommiharibabu — APPROVED** [review]
LGTM.

**Comment by kommiharibabu** [comment]
@quchenhao Can you run a patch build and stress test to confirm that we didn't miss any other scenario that can cause problems?

**Review by hbokhari — APPROVED** [review]
LGTM. I liked reading the comments you have added.

As Hari mentioned, run WT stress tests and MongoDB patch build to verify before merging.

**Comment by quchenhao** [comment]
I don't find any error in the patch build that looks like related to this change. I'll merge.

**Pull request merged**